Fix Blackbox P interval

McGiverGim opened the pull request:
Fixes https://github.com/betaflight/betaflight/issues/5625

The problem is that for some values, the `blackboxGetRateNum(void)` is zero, producing a endless loop in the blackbox explorer because the `P interval` ends with something like `0 / 1`.

If I'm not wrong the `P interval` defines the number of gyro loop cycles that must pass before we capture a Blackbox frame (I or P). Is represented in the Blackbox file in the format `blackboxGetRateNum /  blackboxGetRateDenom` so for example a value of `1 / 7` is that each seven iterations one is an I or P frame.

This change seems to solve the problem, but maybe I've simplified too much. Now with this change the `P interval` is always in the form `1 / blackboxPInterval` and maybe certain rates/configs need a numerator different than 1 but I can't see them in the code. It seems to be always a counter that resets to zero when this value is reached.

You can get more info about this here: https://github.com/betaflight/betaflight/issues/2954

Thoughts?

mikeller commented on src/main/blackbox/blackbox.c:
```diff
@@ -1697,12 +1697,16 @@ int blackboxCalculatePDenom(int rateNum, int rateDenom)
 uint8_t blackboxGetRateNum(void)
 {
-    return blackboxGetRateDenom() * blackboxConfig()->p_denom / blackboxIInterval;
+    return 1;
```
Should be (as far as I can tell):

```
if (blackboxPInterval == 0) {
    return 0;
} else {
￼   return 1;
￼}
```

mikeller commented on src/main/blackbox/blackbox.c:
```diff
@@ -1697,12 +1697,16 @@ int blackboxCalculatePDenom(int rateNum, int rateDenom)
+    if (blackboxPInterval == 0) {
+        return 1;
+    } else {
+        return blackboxPInterval;
```
Should be `return blackboxConfig()->p_denom`.

McGiverGim commented:
@mikeller I'm not too sure about your suggestions. I will try to explain:

The `P denom` is logged in a different blackbox header. Here you see both:
https://github.com/betaflight/betaflight/blob/c8a646ebc3d35b9a6ef1de36190808869fefafe1/src/main/blackbox/blackbox.c#L1198-L1200
The `P denom` is what you say, `a ratio of how many p frames are logged for every I frame` but the problem here is with the `P interval`.
The two functions modified in this PR are only used it the Blackbox header `P interval` (and in some unit tests). They are not used in other parts of the firmware. So they only affect to the Blackbox Explorer interpretation of the data.
Your first suggestions, if I modify the `blackboxGetRateNum()` to let zero be a "possible" value, then we have the problem of the issue. The issue happens because:
```
H P interval:0/1
```

Being zero produces an infinite loop in the Blackbox Explorer because this is always false:
https://github.com/betaflight/blackbox-log-viewer/blob/5eb2a39df1174035065c4795f7921d581a780db2/js/flightlog_parser.js#L861-L865

And sinctactically has no sense to this value to be zero, but I will try to explain this later.

Your second suggestion, use the `return blackboxConfig()->p_denom` in the `blackboxGetRateDenom()` produces this in the blackbox log of the issue:
![image](https://user-images.githubusercontent.com/2673520/38461392-9d27a8e0-3ace-11e8-9566-cde3ef9e012c.png)
My code returns 7 here, your suggestion returns 32. Any value over 7 produces this because the Blackbox thinks that they are missing frames. And of course, the actual unit tests of this function will fail:
https://github.com/betaflight/betaflight/blob/5dc23e662fc9de719bed42781b2447844771d75a/src/test/unit/blackbox_unittest.cc#L289

What is what I think? The `P interval` with a value of `1/7` for example what is saying is that for each 7 iterations of the PID loop, we must write one frame to the blackbox log. This frame can be an I or P frame depending on the P denom, but this is other thing. By this reason it has no sense values like `0/1` as I commented before.

With the actual code, it will always be of the form 1/X because it will never register at rates higher than the PID loop (but I'm not too sure if this is possible). If this is possible then things like `2/1` are possible (each PID loop we must register two frames). The unit tests are all in the form 1/X. I think in the past this other types of values were possible and for this reason the code was complicated. Is possible to register more than one frame for each pid iteration in the actual firmware?

If finally my code is right, I can add some more unit tests with the values of the blackbox files of the issue. We can too remove the `blackboxGetRateNum()` because it is always 1 and hardcode the 1 to the `BLACKBOX_PRINT_HEADER_LINE`.

mikeller commented:
@McGiverGim:

> The two functions modified in this PR are only used in the Blackbox header P interval (and in some unit tests). They are not used in other parts of the firmware.

That is wrong. They are also used in `MSP_BLACKBOX_CONFIG`, and as such it is hard to tell what is depending on them adhering to their original definition.

To my understanding, the original intent of `BLACKBOX_PRINT_HEADER_LINE("P interval", "%d/%d",                   blackboxGetRateNum(), blackboxGetRateDenom());` was to express the P interval in terms of I interval (at least that's the code that you want to change tries to do). I.e. a value of `1/2` would mean that the P interval is half as long as the I interval. My suggested changes try to preserve this. If we want to change this for the P interval to be expressed in terms of PID loop runs then we can, but we'll have to change the blackbox log viewer to match this. You didn't state that this was your goal in this pull request.

> Your first suggestions, if I modify the blackboxGetRateNum() to let zero be a "possible" value, then we have the problem of the issue.

Yes, we need 0 to be possible here, or we need some other way to indicate the special case of `blackboxPInterval == 0`, and no P frames at all being logged, because that is one way to set up logging, and the blackbox log viewer needs to be able to recognise and handle this. Letting P interval be 0 seems to be a practical way to indicate this.

Taking your statements into account, I think it will be best if we abandon 'P denom' in the blackbox log header, and send `blackboxPInterval` as the 'P interval' (both of these are coupled anyway, an there is little value of sending both if we can calculate on from the other), and change the blackbox log viewer to be able to cope with this change.

McGiverGim commented:
@mikeller 
> That is wrong. They are also used in MSP_BLACKBOX_CONFIG, and as such it is hard to tell what is depending on them adhering to their original definition.

Yes, that is true. Is used by Blackbox and can be get by a MSP command. But I'm not too sure if the original definition is what you say. The unit tests were defined some months ago in the https://github.com/betaflight/betaflight/pull/3000 (generated to fix https://github.com/betaflight/betaflight/issues/2954) and they don't adhere to the definition `the P interval in terms of I interval`.  Is more the `P interval` in terms of the loop.
You have the code here:
https://github.com/betaflight/betaflight/blob/247766c30ec3262f733894aa5edf67349ac88a41/src/main/blackbox/blackbox.c#L1715-L1732
You can see than `blackboxIInterval` is calculated in terms of the `gyro.targetLooptime`, and then the `blackboxPInterval` is calculated from the `blackboxIInterval` so indirectly from the `gyro.targetLooptime` too. 
And here you have the unit tests that don't adhere to this definition.
 https://github.com/betaflight/betaflight/blob/5dc23e662fc9de719bed42781b2447844771d75a/src/test/unit/blackbox_unittest.cc#L289
So if we change it, maybe is now when we break things that are depending on because since, at least, july of 2017 they are working in a different manner. Maybe this is a bug introduced then, I don't know, maybe @martinbudden, @ledvinap or @thenickdude (all of them were involved here) can tell us more about what was the original intention.

What you are defining is directly the `p_denom`. If zero, then no p frames. if not, it defines how many frames are captured each 32ms. From this number, 1 is an i frame and the rest are p frames. And the `p_denom` is sended in the Blackbox and MSP command too, so this info is available to blackbox and MSP and I suppose that is what external things are using if needed. I can rename it like the CLI command to `p_ratio`, but I don't think this value must go in the `P interval` header by the reasons explained. But what is sure is that in the Blackbox we need this ratio and I don't think is available in other variables (but can be calculated, but is a risk because some integer rounding).

At the end, I think we have two possibilities:
- The `P interval` of the latest months is wrong, has a bug, and we need to change it as you suggest. I don't know if it will break things (the Blackbox Explorer for sure, others I don't know). In this way the `P interval` and the `P denom` (`P ratio` in CLI) are more or less the same and we need to send a new value to the Blackbox with the ratio frames/loop iterations.
- The `P interval` is "correct" and only has a bug for some special rates that with the rounding of some operations end with values of zero.

mikeller commented:
> But I'm not too sure if the original definition is what you say. The unit tests were defined some months ago in the #3000 (generated to fix #2954) and they don't adhere to the definition the P interval in terms of I interval. Is more the P interval in terms of the loop.

That is not what I said, my statement had nothing to do with `blackboxPInterval`, but with 'P Interval' as sent in `BLACKBOX_PRINT_HEADER_LINE("P interval", "%d/%d", blackboxGetRateNum(), blackboxGetRateDenom())`: 

> To my understanding, the original intent of BLACKBOX_PRINT_HEADER_LINE("P interval", "%d/%d", blackboxGetRateNum(), blackboxGetRateDenom()); was to express the P interval in terms of I interval (at least that's the code that you want to change tries to do). I.e. a value of 1/2 would mean that the P interval is half as long as the I interval.

If you look at the implementation of `blackboxGetRateNum()` and `blackboxGetRateDenom(), it appears that this was expressed as a fraction of the 'I interval', as I stated above, and not as a number of time slots / PID loop times. Changing this in a speculative way to diverge from this definition, and massaging the returned values by discarding information (`blackboxPInterval == 0` and `blackboxPInterval == 1` result in the same value being set in the header) does not seem to me to be a good approach to solving this.

McGiverGim commented:
Yes, I'm with you in the latest part: my solution is not 100% correct. It is a lot more simple than the original, and I think that if my solution were correct, the others developers will have implemented it in this way from the beginning, so there is something that I don't see. It works for the actual unit tests, and works to solve the problem with some blackbox logs attached to the issue, but there must be some case that maybe it does not solve.

I only wanted to "reiterate" that, a value of `Blackbox P interval` of `1/32` is not correct for `p_ratio=32`, or at least this will break the unit tests and the behaviour of latest Betaflight versions. My solution, before simplify, is more like `Blackbox P interval = Var I interval / Var p_ratio`, but in the end, after simplify, it is `Blackbox P interval = 1 / Var P interval` that is more or less what you are saying. Maybe we were talking about more or less the same?

mikeller commented:
> It is a lot more simple than the original, and I think that if my solution were correct, the others developers will have implemented it in this way from the beginning, so there is something that I don't see.

I think the reason it is like it is now is that this was a gradual transformation from the original blackbox implementation that just had a denominator to determine the number of frames to be logged, and then calculated the frame types to be used based on this, to the new system that has a fixed frequency for I frames, and then uses a ratio to determine how many P frames to log inbetween. The logging part was done, but the part where the configuration settings are stored in the header file was never updated to reflect this, and because the two approaches are quite similar, it mostly worked, for most of the cases.

And yes, I think we are talking about sending the same values. In my initial comments I was fooled by the misleading name `p_denom`, This should either be `1 / blackboxPInterval`, or `1 / p_denom`. But really we should just get rid of the fraction for 'P interval', and send `blackboxPInterval` for it.

McGiverGim commented:
This has sense to me. So if I'm not wrong, for example, this is one of the blackbox files that gave problems:
```
H I interval:253
H P interval:0/1
H P denom:32
H looptime:126
```
If I'm not wrong, you are saying that we must send:
```
H I interval:253
H P interval:7
H P ratio:32
H looptime:126
```
The only changes are:
- The `P interval`, that gets rid of the numerator and all the complicated calculations and sends only the `blackboxPInterval`. In this way a `H P interval:0` is a valid value to reflect there are not P frames (only I frames).
- And the `P denom` that changes the name to `P ratio`.

This is clear, and it reflects more the actual behaviour of the blackbox explorer. 

Is that correct? If it is, I will change this PR and change the Blackbox Explorer to reflect this.

McGiverGim commented:
One thing more, if we do this change... What must we do with the MSP command? Maintain it as now with X/Y format?

mikeller commented:
Yes, that is what I am proposing. 'P interval' of 0 is still a special case, but having 'P ratio' as 0 at the same time gives an indication of what it means.

I don't think MSP is sending anything as a fraction, it's just sending  the individual values. We should set the ones we don't calculate any more to 0, and let clients deal with it when they update to support MSP API 1.39.

McGiverGim commented:
Maybe is more "compatible" to return a 1 and not a zero (the vast majority of them will work and will be transparent). Something like:
>sbufWriteU8(dst, blackboxGetRateNum());
>sbufWriteU8(dst, blackboxGetRateDenom());

to
>sbufWriteU8(dst, 1);
>sbufWriteU8(dst, blackboxGetRateDenom());

With this, the 99% of the cases will work because the `1/blackboxGetRateDenom()` contains the correct value for the old protocol (except when `P interval` is zero, but in this case, the `blackboxGetRateDenom() is zero so it is not too bad.

McGiverGim commented:
By the moment, I've modified it sending zero in the MSP as you suggest, but maybe I've gone too far in the renaming of `p_denom` by `p_ratio`... 

- I've renamed the Blackbox header, as we discuss
- I've renamed the variable too, including the PG
- I've renamed the CMS input (and the variables that reference it)

I've added too two more test unit cases with the ones that gave problems in the issue.

mikeller commented on src/main/cms/cms_menu_blackbox.c:
```diff
@@ -195,7 +195,7 @@ static OSD_Entry cmsx_menuBlackboxEntries[] =
     { "(USED)",      OME_String,  NULL,            &cmsx_BlackboxDeviceStorageUsed,                           0 },
     { "(FREE)",      OME_String,  NULL,            &cmsx_BlackboxDeviceStorageFree,                           0 },
-    { "P DENOM",     OME_UINT16,  NULL,            &(OSD_UINT16_t){ &blackboxConfig_p_denom, 1, INT16_MAX, 1 },0 },
+    { "P DENOM",     OME_UINT16,  NULL,            &(OSD_UINT16_t){ &blackboxConfig_p_ratio, 1, INT16_MAX, 1 },0 },
```
We should change the title here to `P RATIO` as well.

McGiverGim commented:
Changes done as requested. I will push an update with this changes to the Blackbox Explorer too asap.

What will be the version target for this? 3.4.0 or 3.3.3?

mikeller commented:
Good question re 3.3 or 3.4. It is a bugfix, but since it will need a new release of the blackbox log viewer to work I am not sure about putting it into 3.3. If so, it should go into 3.3.2, which has an increment of the MSP API version already, so clients using the changed values can use the API version to distinguish the two formats.
Is there more outstanding work to do for a release of the blackbox log viewer, or will we be able to release it in time for 3.3.2?

McGiverGim commented:
Changes to the Blackbox ready. It does not need the version, it has been easier to see if it is in the form  X/Y or not. The new `P ratio` (or the old P denom) was not used at all. 

I think the Blackbox is ready, but I don't know if @basdelfos has something pending for it.

mikeller commented:
3.3 it is then!

basdelfos commented:
> I think the Blackbox is ready, but I don't know if @basdelfos has something pending for it.

Nope, all good.

mikeller commented:
Should we look into releasing 3.0.0 for Windows, MacOS, linux, and the Chrome web store then soon-ish, guys?

McGiverGim commented:
Yes, I think is a good moment for it, since august of 2017 :P
